frontend/src/components/AnnotationRenderer.vue

```diff
@@ -46,6 +46,62 @@ import {DocumentType} from '@/enum/DocumentTypes';
 import MarkdownRenderer from "@/components/MarkdownRenderer.vue";
 import _ from "lodash"
 import {getValueFromKeyPath} from "@/utils/dict";
+import {compile, registerPlugin} from "jse-eval";
+
+// set up jsep extensions
+// arrow functions, to allow things like "annotations.myCheckboxes.some(v => v < 4)"
+// to test multi-select checkboxes
+import jsepArrow from "@jsep-plugin/arrow";
+// new operator to allow for date comparisons with "new Date()" or similar
+import jsepNew from "@jsep-plugin/new";
+// hex and octal numeric literals
+import jsepNumbers from "@jsep-plugin/numbers";
+// object expressions
+import jsepObject from "@jsep-plugin/object";
+// regex support, to allow for if tests that check a text field against a regex
+import jsepRegex from "@jsep-plugin/regex";
+// destructuring, useful in filtering functions etc.
+import jsepSpread from "@jsep-plugin/spread";
+
+registerPlugin(
+    jsepArrow,
+    jsepNew,
+    jsepNumbers,
+    jsepObject,
+    jsepRegex,
+    jsepSpread,
+)
+
+// Additional global objects that will be available to if expressions
+const EXPR_CONTEXT_PROTO = Object.freeze(Object.assign(Object.create(null), {
+  String, Array, RegExp,
+  JSON, Math, parseInt, parseFloat,
+  encodeURIComponent, decodeURIComponent,
+
+  // for Object we include only side-effect-free methods that are
+  // likely to be needed in expressions.  In particular we do *not*
+  // expose Object.assign or Object.getPrototypeOf, as those allow
+  // expressions to manipulate the document and annotation objects
+  // or even pollute the global prototypes of Function, Array, Boolean,
+  // Number, String or Object.  The jse-eval library prevents access
+  // to any property named prototype, __proto__ or constructor, but
+  // that can be circumvented by using a construct like
+  //
+  // Object.assign(Object.getPrototypeOf(()=>0), {toString: () => 'this is evil'}),
+  //
+  // Forbidding Object.assign and getPrototypeOf closes this loophole.
+  // This doesn't avoid all possible side effects (e.g. you could do
+  // document.someArrayProperty.push('foo') or similar) but it gets rid of
+  // the most serious hole.
+  Object: Object.freeze(Object.assign(Object.create(null), {
+    entries: Object.entries,
+    fromEntries: Object.fromEntries,
+    keys: Object.keys,
+    values: Object.values,
+    getOwnPropertyNames: Object.getOwnPropertyNames,
+    hasOwn: Object.hasOwn,
+  }))
+}))
```

Comment:
This is the bit that has been bugging me slightly, and which I want to resolve one way or another before this PR is ready to merge.  Any time when we start accepting and evaluating arbitrary JS expressions from the (manager) user we need to think about security, and it's tricky to find the right balance between locking down things that are dangerous but still making the feature useful.

The expression parser is configured to reject assignment expressions but I realised that allowing `Object.assign` would bypass that.  And it blocks access to `prototype`, `__proto__` and `constructor` properties but `Object.getPrototypeOf` would bypass that...

The solution I've adopted here is to allow-list the static `Object.x` methods that are visible to `if` expressions, is that sufficient or do we need to do something similar with `Array` and `String`?  Would it be safer to block function expressions altogether, though they arere useful for things like `myArray.every((v) => v.something())`...

I don't give the expressions access to things like the global `window` object, but given `annotation` and `document` are not just plain objects but are enhanced with Vue's getters and setters for scope watching, does that provide a backdoor way to get at the DOM element that the component is attached to, and thence to things like `window.alert`?

Basically how paranoid are we/should we be?

Reply:
> given `annotation` and `document` are not just plain objects but are enhanced with Vue's getters and setters for scope watching, does that provide a backdoor way to get at the DOM element that the component is attached to, and thence to things like `window.alert`?

Why yes, yes it does...

```javascript
annotation.__ob__.dep.subs[0].vm.$el.ownerDocument.defaultView.alert('haha')
```